File: leetcode/347. Top K Frequent Elements.cpp
struct FreqNum{
    int freq;
    int num;
};

struct cmp{
    bool operator()(FreqNum a, FreqNum b){
        return a.freq > b.freq;
    }
};

class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> dic;
        for(const auto num: nums){
            ++dic[num];
        }
        
        priority_queue<FreqNum, vector<FreqNum>, cmp> pq;
        for(const auto& iter: dic){
            FreqNum item;
            item.freq = iter.second;
            item.num = iter.first;
            pq.push(move(item));
            if(pq.size() > k){
                pq.pop();
            }
        }
        
        vector<int> res;
        while(!pq.empty()){
            res.push_back(pq.top().num);
            pq.pop();
        }
        reverse(res.begin(), res.end());
        return res;
    }
};

File: leetcode/347. Top K Frequent Elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        dic = defaultdict(int)
        max_freq = 0
        for num in nums:
            dic[num] += 1
            max_freq = max(max_freq, dic[num])
        
        lst = [(freq, num) for num, freq in dic.items()]
        buckets = [[] for _ in range(max_freq+1)]

        for num, freq in dic.items():
            buckets[freq].append(num)

        res = []
        for i in range(len(buckets)-1, -1, -1):
            lst = buckets[i]
            res += lst[:min(k-len(res), len(lst))]
            if len(res) == k:
                break
        return res
        
        
  
  
from heapq import heappush, heappop

class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        dic = {}
        for num in nums:
            if num not in dic:
                dic[num] = 0
            dic[num] += 1
            
        freqs = []
        for num, freq in dic.items():
            if len(freqs) == k:
                if freqs[0][0] < freq:
                    heappop(freqs)
                    heappush(freqs, (freq, num))
            else:
                heappush(freqs, (freq, num))
        return [num for _, num in freqs]
